display_list/display_list_complexity_helper.h

```diff
@@ -0,0 +1,275 @@
+// found in the LICENSE file.
+
+#ifndef FLUTTER_FLOW_DISPLAY_LIST_COMPLEXITY_HELPER_H_
+#define FLUTTER_FLOW_DISPLAY_LIST_COMPLEXITY_HELPER_H_
+
+#include "flutter/display_list/display_list_complexity.h"
+#include "flutter/display_list/display_list_dispatcher.h"
+#include "flutter/display_list/display_list_utils.h"
+
+namespace flutter {
+
+// The Metal and OpenGL complexity calculators use benchmark data gathered
+// using the display_list_benchmarks test suite on real devices.
+//
+// Constants of proportionality and trends were chosen to better match
+// larger numbers rather than smaller ones. This may turn out to be the
+// wrong decision, but the rationale here is that with the smaller numbers,
+// we have:
+//
+// a) More noise
+// b) Less absolute difference. If the constant we've chosen is out by 50%
+//    on a measurement that is 0.001ms, that's less of an issue than if
+//    the measurement is 15ms.
+// c) Smaller numbers affect the caching decision negligibly; the caching
+//    decision is likely to be driven by slower ops rather than faster ones.
+//
+// In some cases, a cost penalty is used to figure out the cost of an
+// attribute such as anti-aliasing or fill style. In some of these, the
+// penalty is proportional to some other value such as the radius of
+// a circle. In these cases, we ensure that the penalty is never smaller
+// than 1.0f.
+//
+// The error bars in measurement are likely quite large and will
+// vary from device to device, so the goal here is not to provide a
+// perfectly accurate representation of a given DisplayList's
+// complexity but rather a very rough estimate that improves upon our
+// previous cache admission policy (op_count > 5).
+//
+// There will likely be future work that will improve upon the figures
+// in here. Notably, we do not take matrices or clips into account yet.
+//
+// The scoring is based around a baseline score of 100 being roughly
+// equivalent to 0.0005ms of time. With a 32-bit unsigned integer, this
+// would set the maximum time estimate associated with the complexity score
+// at about 21 seconds, which far exceeds what we would ever expect a
+// DisplayList to take rasterising.
+//
+// Finally, care has been taken to keep the computations as cheap as possible.
+// We need to be able to calculate the complexity as quickly as possible
+// so that we don't end up wasting too much time figuring out if something
+// should be cached or not and eat into the time we could have just spent
+// rasterising the DisplayList.
+//
+// In order to keep the computations cheap, the following tradeoffs were made:
+//
+// a) Limit all computations to simple division, multiplication, addition
+//    and subtraction.
+// b) Try and use integer arithmetic as much as possible.
+// c) If a specific draw op is logarithmic in complexity, do a best fit
+//    onto a linear equation within the range we expect to see the variables
+//    fall within.
```

Comment:
Can you suggest an approach in this comment for adding a new DL op?

Reply:
Done